common/hyperscaler/rules/parsing_results.go

```diff
@@ -0,0 +1,112 @@
+package rules
+
+import "fmt"
+
+type ParsingResults struct {
+	resolvedRules map[string]*ParsingResult
+	uniquenessSet map[string]*ParsingResult
+	uniqueResults []*ParsingResult
+
+	Results []*ParsingResult
+}
+
+func (p *ParsingResults) Apply(entry string, rule *Rule, err error) {
+	result := NewParsingResult(entry, rule)
+
+	if err != nil {
+		result.AddParsingError(err)
+	} else {
+		if rule.IsResolved() {
+			p.resolvedRules[rule.StringNoLabels()] = result
+		}
+	}
+
+	p.Results = append(p.Results, result)
+}
+
+func NewParsingResults() *ParsingResults {
+	return &ParsingResults{
+		Results:       make([]*ParsingResult, 0),
+		resolvedRules: make(map[string]*ParsingResult),
+		uniquenessSet: make(map[string]*ParsingResult),
+		uniqueResults: make([]*ParsingResult, 0),
+	}
+}
+
+func (p *ParsingResults) IsResolved(resolvingKey string) bool {
+	_, resolvingRuleExists := p.resolvedRules[resolvingKey]
+	return resolvingRuleExists
+}
+
+func (p *ParsingResults) CheckUniqueness() {
+
+	for _, result := range p.Results {
+
+		if result.HasErrors() {
+			p.uniqueResults = append(p.uniqueResults, result)
+			continue
+		}
+
+		key := result.Rule.SignatureWithValues()
+
+		alreadyExists := false
+		var item *ParsingResult
+		item, alreadyExists = p.uniquenessSet[key]
+
+		if !alreadyExists {
+
+			p.uniquenessSet[key] = result
+			p.uniqueResults = append(p.uniqueResults, result)
+
+		} else {
+
+			err := fmt.Errorf("Duplicated rule with previously defined rule: '%s'", item.Rule.StringNoLabels())
+
+			result.AddProcessingError(err)
+
+			p.uniqueResults = append(p.uniqueResults, result)
+
+		}
+	}
+
+	p.Results = p.uniqueResults
+}
+
+func (p *ParsingResults) CheckSignatures() {
+	uniqueResults := make([]*ParsingResult, 0, len(p.Results))
+	proposedResolutions := make(map[string]bool)
+	signatureSet := NewSignatureSet(p.Results)
+
+	for _, result := range p.Results {
+
+		if result.HasErrors() {
+			uniqueResults = append(uniqueResults, result)
```

Comment:
We append result to uniqueResults in each path of execution inside for loop over p.Results.